### PR TITLE
Adding the possibility to use pdftotext options

### PR DESCRIPTION
**Pull request by nyamsprod**
- the following method is added `Pdf::setOptions`
- `Pdf::getText` gained a third parameters to allow specifying optional options
- Adding `InvalidOption` exception when the added options is malformed.

An option is considered valid if it is a string which starts with a "-" hyphen.

### CODE REVIEW

File: src/Pdf.php [hunk 1]
```diff
@@ -12,38 +13,64 @@ class Pdf
...
         $this->binPath = $binPath ?? '/usr/bin/pdftotext';
     }
 
-    public function setPdf(string $pdf) : Pdf
+    public function setPdf(string $pdf) : self
     {
-        if (!file_exists($pdf)) {
-            throw new PdfNotFound("could not find pdf {$pdf}");
+        if (!\is_readable($pdf)) {
+            throw new PdfNotFound(\sprintf('could not find or read pdf `%s`', $pdf));
```
Comment (freekmurze) [inline src/Pdf.php]: I preferred the old way of formatted the exception message, could you revert this change?

File: src/Pdf.php [hunk 2]
```diff
@@ -12,38 +13,64 @@ class Pdf
...
     {
         $this->binPath = $binPath ?? '/usr/bin/pdftotext';
     }
 
-    public function setPdf(string $pdf) : Pdf
+    public function setPdf(string $pdf) : self
     {
-        if (!file_exists($pdf)) {
-            throw new PdfNotFound("could not find pdf {$pdf}");
+        if (!\is_readable($pdf)) {
```
Comment (freekmurze) [inline src/Pdf.php]: I'm not to keen on that `\` for ensuring the global functions are used. Could you remove all `\` prefixes you've used in function calls?

File: README.md [hunk 3]
```diff
@@ -78,7 +79,22 @@ $text = (new Pdf('/custom/path/to/pdftotext'))
...
- \Spatie\PdfToText\Pdf::getText('book.pdf', '/custom/path/to/pdftotext')
+ \Spatie\PdfToText\Pdf::getText('book.pdf', '/custom/path/to/pdftotext');
+```
+
+Sometimes you may want to use [pdftotext options](https://linux.die.net/man/1/pdftotext). To do so you can set them up using the `setOptions` method
+
+```php
+$text = (new Pdf())
+    ->setPdf('table.pdf')
+    ->setOptions(['-layout'])
```
Comment (freekmurze) [inline README.md]: Could you make this work so the `-` isn't needed?

File: src/Pdf.php [hunk 4]
```diff
@@ -12,38 +13,64 @@ class Pdf
...
         }
 
         $this->pdf = $pdf;
 
         return $this;
     }
 
+    public function setOptions(array $options) : self
+    {
+        $this->options = \array_map([$this, 'filterOption'], $options);
```
Comment (freekmurze) [inline src/Pdf.php]: The filtering of the options is not needed. Let's make it so that the users doesn't have to prefix options with `-`

Comment (nyamsprod) [reply src/Pdf.php]: I added the option to add or not the starting hyphen .. if present it won't be added if not it will be added. the bonus is that the array is self validated by PHP so if you an array which contains something which is not stringable it will throw a `TypeError`.

File: README.md [hunk 5]
```diff
@@ -65,11 +65,12 @@ $text = (new Pdf())
 Or easier:
 
 ```php
- \Spatie\PdfToText\Pdf::getText('book.pdf')
+ \Spatie\PdfToText\Pdf::getText('book.pdf');
 ```
 
 By default the package will assume that the `pdftotext` is located at `/usr/bin/pdftotext`.
-If you're using the a different location pass the path to the binary in constructor
+If you're command is located at a different location pass its binary path to constructor
```
Comment (m1guelpf) [inline README.md]: If your

File: src/Pdf.php [hunk 6]
```diff
@@ -12,38 +13,63 @@ class Pdf
...
+
     public function text() : string
     {
-        $process = new Process("{$this->binPath} " . escapeshellarg($this->pdf) . " -");
-        $process->run();
+        $arguments = $this->options;
+        $arguments[] = escapeshellarg($this->pdf);
+        $arguments[] = '-';
 
+        $commandline = $this->binPath.' '.implode(' ', $arguments);
```
Comment (freekmurze) [inline src/Pdf.php]: Would it be possible to use and array with the arguments here, instead of concatenating to a string?

See: https://symfony.com/doc/current/components/process.html

Comment (nyamsprod) [reply src/Pdf.php]: @freekmurze that's not possible because of your travis settings. You minimum requirement for symfony/process do not allow the use of array arguments 😢 unless you update the project requirements which means BC break in my book.

**Comment by nyamsprod** [comment]
@freekmurze The thing is I did not want to change the library requirements... but if it's okay for you what is your prefer target ? v3.3 or v4.0 ?

**Comment by nyamsprod** [comment]
@freekmurze 

When using array parameters I get an exception with this simple snippet

```php
<?php

use Symfony\Component\Process\Process;

require 'vendor/autoload.php';

$arguments = [
    '/usr/bin/pdftotext',
    '-layout',
    '-r 72',
    '/path/to/scoreboard.pdf',
    '-',
];

echo '<pre>', PHP_EOL;
$process = new Process($arguments);
echo $process->getCommandLine(), PHP_EOL;
$process->mustRun();
```
Output

```bash
'/usr/bin/pdftotext' '-layout' '-r 72' '/path/to/scoreboard.pdf' '-'
The command "'/usr/bin/pdftotext' '-layout' '-r 72' '/path/to/scoreboard.pdf' '-'" failed.

Exit Code: 99(Unknown error)
```

If you have an idea as to why ... I'll update the PR 👍

**Comment by freekmurze** [comment]
>  v3.3 or v4.0

I'm thinking we could both versions.

> If you have an idea as to why 

I don't see the problem atm. What happens if you run that command from the error message in your console? Might give a more detailed error description.

**Comment by nyamsprod** [comment]
so the issue is around the use of value around options the following command works:

`/usr/bin/pdftotext '-layout' '-r' 96 'tests/testfiles/scoreboard.pdf' '-'` 

while the command below does not work

`/usr/bin/pdftotext '-layout' '-r 96' 'tests/testfiles/scoreboard.pdf' '-'`

notice the quoting difference around the `-r` option value

**Comment by nyamsprod** [comment]
I think I found the solution ... 👍

File: README.md [hunk 7]
```diff
@@ -10,9 +10,12 @@
 This package provides a class to extract text from a pdf.
 
 ```php
- \Spatie\PdfToText\Pdf::getText('book.pdf'); //returns the text from the pdf
-```
+<?php
```
Comment (freekmurze) [inline README.md]: Remove all the  `<?php ` tags you added. They just take in unnecessary space.

File: README.md [hunk 8]
```diff
@@ -27,27 +30,29 @@ The best postcards will get published on the open source page on our website.
 ## Requirements
 
 Behind the scenes this package leverages [pdftotext](https://en.wikipedia.org/wiki/Pdftotext). You can verify if the binary installed on your system by issueing this command:
-```
-which pdftotext
+
+```bash
+$ which pdftotext
```
Comment (freekmurze) [inline README.md]: Remove these `$` you added, they make it harder to copy and paste the line.

File: README.md [hunk 9]
```diff
@@ -57,6 +62,10 @@ $ composer require spatie/pdf-to-text
 Extracting text from a pdf is easy.
 
 ```php
+<?php
+
+use Spatie\PdfToText\Pdf;
```
Comment (freekmurze) [inline README.md]: Remove all `use` statements, except the first one at the top of the readme.

File: README.md [hunk 10]
```diff
@@ -65,29 +74,67 @@ $text = (new Pdf())
...
+
+echo Pdf::getText('book.pdf', null, ['layout', 'opw myP1$$Word']);
 ```
 
 ## Change log
 
-Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.
+Please see [CHANGELOG](CHANGELOG.md) for more informations on what has changed recently.
```
Comment (freekmurze) [inline README.md]: This is not correct. Could you revert the change?

**Comment by freekmurze** [comment]
Thanks!
